### PR TITLE
Updated Store interface

### LINKED ISSUE #66: Extend/refactor BlockStore interface and implementations
Currently, [storage is defined only for blocks](https://github.com/lazyledger/optimint/blob/main/store/blockstore.go). We need to also store Commits.

In Tendermint `BlockStore` is responsible also for commits. We can follow this approach or just rename `BlockStore` to something more generic and add methods for Commits.

### PR DESCRIPTION
**Pull request by tzdybal**
This PR adds support for `Commit` saving and loading in `Store`. Additionally it renames `BlockStore` to just `Store` and add some other refactorings (method extraction, etc).

Resolves #66.

### CODE REVIEW

File: store/store.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
+package store
+
+import (
+	"encoding/binary"
+	"errors"
+	"sync"
+
+	"go.uber.org/multierr"
```
Comment (adlerjohn) [inline store/store.go:8]: 👀

I'm not sure how I feel about relying on a package from Uber. What's their track record on maintaining?

File: store/store.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+func New() Store {
+	return &DefaultStore{db: NewInMemoryKVStore()}
+}
+
+func (s *DefaultStore) Height() uint64 {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	return s.height
+}
```
Comment (adlerjohn) [inline store/store.go:39]: Some of these method aren't super-obvious, can we get some docstrings for exported methods?

File: store/store.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	return append(blockPrefix[:], hash[:]...)
+}
+
+func getCommitKey(hash [32]byte) []byte {
+	return append(commitPrefix[:], hash[:]...)
+}
+
+func getIndexKey(height uint64) []byte {
+	buf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(buf, height)
```
Comment (adlerjohn) [inline store/store.go:140]: Is there a reason to use LE instead of BE here?

Comment (tzdybal) [reply store/store.go:140]: I thought that LE is more common on most popular architectures (see https://en.wikipedia.org/wiki/Endianness#Current_architectures). TBH I was not thinking that about this - I just needed _some_ way to store uint64 in []byte.

File: store/store.go (hunk 1 above)
Comment (tzdybal) [reply store/store.go:8]: I would say that sentiment for uber go modules is very good in go community.

Last commit for this particular package was committed 15 days ago.

File: store/store.go (hunk 3 above)
Comment (adlerjohn) [reply store/store.go:140]: My preference would be BE, so as to be inline with https://github.com/lazyledger/smt/issues/20

File: store/store.go (hunk 2 above)
Comment (tzdybal) [reply store/store.go:39]: Commented on interface level.

File: store/store.go (hunk 3 above)
Comment (liamsi) [reply store/store.go]: resolved in fa0aa80

File: store/store.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	commitPrefix = [1]byte{3}
+)
+
+type DefaultStore struct {
+	db KVStore
+
+	height uint64
+
+	// mtx protects height
+	mtx sync.RWMutex
```
Comment (Wondertan) [inline store/store.go:26]: nit: keeping the value and it's guardian close to each other increases readibility

File: store/store.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+var _ Store = &DefaultStore{}
+
+func New() Store {
+	return &DefaultStore{db: NewInMemoryKVStore()}
+}
```
Comment (Wondertan) [inline store/store.go:33]: So it is expected to have only in mem version of Store?

File: store/store.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+func (s *DefaultStore) SaveBlock(block *types.Block, commit *types.Commit) error {
+	hash, err := hash.Hash(&block.Header)
+	if err != nil {
+		return err
+	}
+	blockKey := getBlockKey(hash)
+	commitKey := getCommitKey(hash)
+	indexKey := getIndexKey(block.Header.Height)
```
Comment (Wondertan) [inline store/store.go:48]: nit: Maybe define these before code that specifically uses those vars? Just below locks/

File: store/store.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	commitBlob, err := commit.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	// TODO(tzdybal): use transaction for consistency of DB
```
Comment (Wondertan) [inline store/store.go:62]: I feel the importance of creating an issue for that.

File: store/store.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+	err = multierr.Append(err, s.db.Set(blockKey, blockBlob))
+	err = multierr.Append(err, s.db.Set(commitKey, commitBlob))
+	err = multierr.Append(err, s.db.Set(indexKey, hash[:]))
+
+	if block.Header.Height > s.height {
+		s.height = block.Header.Height
+	}
+
```
Comment (Wondertan) [inline store/store.go:70]: So we set higher height even if we failed to write to DB? Is it intentional?

File: store/store.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+		return err
+	}
+
+	commitBlob, err := commit.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
```
Comment (Wondertan) [inline store/store.go:61]: Why we also lock for DB writes and not just for height change?

File: store/store.go (hunk 7 above)
Comment (Wondertan) [reply store/store.go:62]: In case we are using not just in mem db

**Review by Wondertan — COMMENTED** [review]
mostly nits and questions

File: store/types.go [hunk 10]
```diff
@@ -2,11 +2,21 @@ package store
...
 import "github.com/lazyledger/optimint/types"
 
-type BlockStore interface {
+// Store is minimal interface for storing and retrieving blocks and commits.
+type Store interface {
+	// Height returns height of the highest block in store.
 	Height() uint64
 
-	SaveBlock(block *types.Block) error
+	// SaveBlock saves block along with it's commit.
```
[the remark this answers is not in the document]
Comment (tzdybal) [reply store/types.go:10]: Good point!

File: store/store.go (hunk 5 above)
Comment (tzdybal) [reply store/store.go:33]: No, it's the only implementation right now ;) See: #67.

File: store/store.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
+package store
```
[the remark this answers is not in the document]
Comment (tzdybal) [reply store/store.go:1]: Github didn't peek the renaming from blockstore.go to store.go. This is strange.

File: store/store.go (hunk 7 above)
Comment (tzdybal) [reply store/store.go:62]: Created issue for this: #80.

File: store/store.go (hunk 8 above)
Comment (tzdybal) [reply store/store.go:70]: Good catch, it's not intentional.

File: store/store.go (hunk 4 above)
Comment (tzdybal) [reply store/store.go:26]: The comment is misleading. Mutex is used to ensure that `height` is in sync with `db`.

File: store/store.go (hunk 9 above)
Comment (tzdybal) [reply store/store.go:57]: To ensure that height is in sync with db.

**Review by liamsi — APPROVED** [review]
lgtm

**Pull request merged**